### PR TITLE
Add nodename feature

### PR DESCRIPTION
**Pull request by hwware**
This PR adds the feature of human-readable nodenames to clusters. In the PR, by default, the nodename will be set to ip_port. Specific nodenames can be set by using the command "CLUSTER SETNAME" and can be viewed by using "CLUSTER NODES" or "CLUSTER MYNAME". The nodename is gossiped using the gossip channel so if a nodename is assigned or changed then the other nodes can update the name accordingly. 
Examples of the commands is shown below:
image

The cluster lookup is modified to use the human readable name to find the node is the node id is not provided, an example is shown below:
image

### CODE REVIEW

File: src/redis-cli.c [hunk 1]
```diff
@@ -4006,24 +4006,28 @@ static int clusterManagerNodeLoadInfo(clusterManagerNode *node, int opts,
         lines = p + 1;
         char *name = NULL, *addr = NULL, *flags = NULL, *master_id = NULL,
              *ping_sent = NULL, *ping_recv = NULL, *config_epoch = NULL,
-             *link_status = NULL;
+             *link_status = NULL, *hname = NULL, *custom_name = NULL;
```
Comment (hwware) [inline src/redis-cli.c]: custom_name is an integral type, do we need set it as NULL? i prefer using  custom_name = 0

File: src/redis-cli.c [hunk 2]
```diff
@@ -4006,24 +4006,28 @@ static int clusterManagerNodeLoadInfo(clusterManagerNode *node, int opts,
...
         UNUSED(link_status);
+        UNUSED(hname);
+        UNUSED(custom_name);
```
Comment (hwware) [inline src/redis-cli.c]: I think we could remove this line

File: src/redis-cli.c [hunk 3]
```diff
@@ -4006,24 +4006,28 @@ static int clusterManagerNodeLoadInfo(clusterManagerNode *node, int opts,
...
-            case 3: master_id = token; break;
-            case 4: ping_sent = token; break;
-            case 5: ping_recv = token; break;
-            case 6: config_epoch = token; break;
-            case 7: link_status = token; break;
+            case 1: hname = token; break;
+            case 2: addr = token; break;
+            case 3: flags = token; break;
+            case 4: master_id = token; break;
+            case 5: custom_name = token; break;
```
Comment (hwware) [inline src/redis-cli.c]: If we have some risks for this? a char pointer point to an integral

File: src/cluster.h [hunk 4]
```diff
@@ -6,9 +6,10 @@
  *----------------------------------------------------------------------------*/
 
 #define CLUSTER_SLOTS 16384
-#define CLUSTER_OK 0            /* Everything looks ok */
-#define CLUSTER_FAIL 1          /* The cluster can't work */
-#define CLUSTER_NAMELEN 40      /* sha1 hex length */
+#define CLUSTER_OK 0          /* Everything looks ok */
+#define CLUSTER_FAIL 1        /* The cluster can't work */
+#define CLUSTER_NAMELEN 40    /* sha1 hex length */
+#define CLUSTER_HUMANNAMELEN 32    /* Max length of human readable node name */
```
Comment (hwware) [inline src/cluster.h]: I think CLUSTER_HUMAN_NAMELEN is more readable, how do you think?

File: src/redis-cli.c (hunk 2 above)
Comment (hwware) [reply src/redis-cli.c]: This is to avoid the warning of unused variable

File: src/redis-cli.c (hunk 1 above)
Comment (daniel-house) [reply src/redis-cli.c]: `char ... *custom_name = NULL;` is correct.  It states that `custom_name` is the address of a character, which usually means that later you will set it to the start of a string.  `NULL` is just 0, so it means the same to the compiler, but `NULL` tells the human reading the code that the coder is aware that they are assigning an address to `custom_name`.

File: src/cluster.h (hunk 4 above)
Comment (daniel-house) [reply src/cluster.h]: For macros it is best to use `_`  between every word.  I would name it `CLUSTER_HUMAN_NAME_LEN`, but it isn't a big deal.

Comment (daniel-house) [reply src/cluster.h]: Why were lines 9, 10 and 11 changed?

File: src/cluster.h [hunk 5]
```diff
@@ -113,6 +114,7 @@ typedef struct clusterNodeFailReport {
 typedef struct clusterNode {
     mstime_t ctime; /* Node object creation time. */
     char name[CLUSTER_NAMELEN]; /* Node name, hex string, sha1-size */
+    char hname[CLUSTER_HUMANNAMELEN]; /* Human readable name for node */
```
Comment (daniel-house) [inline src/cluster.h]: If you want to allow the name to be `CLUSTER_HUMANNAMELEN` characters long, then this declaration needs to be
~~~
char hname[CLUSTER_HUMANNAMELEN + 1]
~~~
to allow for the null terminator.  A buffer overrun caused by `strcpy` appending the null terminator is one of the most common ways for C code to crash intermittently.

File: src/cluster.h [hunk 6]
```diff
@@ -206,12 +209,14 @@ typedef struct {
...
-    uint16_t flags;             /* node->flags copy */
-    uint16_t pport;             /* plaintext-port, when base port is TLS */
+    char ip[NET_IP_STR_LEN];            /* IP address last time it was seen */
+    char hname[CLUSTER_HUMANNAMELEN];   /* Human readable name for node */
+    uint16_t port;                      /* base port last time it was seen */
+    uint16_t cport;                     /* cluster port last time it was seen */
+    uint16_t flags;                     /* node->flags copy */
+    uint16_t pport;                     /* plaintext-port, when base port is TLS */
     uint16_t notused1;
+    int custom_name;                    /* Indicates whether custom nodename is used */
```
Comment (daniel-house) [inline src/cluster.h]: Sometimes `custom_name` is `int`, sometimes it is `char *`.  This is very confusing.  When you are creating a flag like this one I recommend a name like `custom_name_defined` or `has_custom_name` and a comment on the `char *` like `meaningless when custom_name_defined is false.  I see that `custom_name` is `int` in the CLI code, and `char *` in the server data structures, which reduces the risk of confusion, but I also see from the comments on this PR that the confusion has occurred.

File: src/cluster.h [hunk 7]
```diff
@@ -139,6 +141,7 @@ typedef struct clusterNode {
     int cport;                  /* Latest known cluster port of this node. */
     clusterLink *link;          /* TCP/IP link with this node */
     list *fail_reports;         /* List of nodes signaling this as failing */
+    int custom_name;
```
Comment (daniel-house) [inline src/cluster.h]: This line should have a comment describing what `custom_name` means.  I recommend changing the field name to `has_human_readable_name` and putting the two fields next to each other in the data structure.

File: src/cluster.h (hunk 5 above)
Comment (daniel-house) [reply src/cluster.h]: I recommend changing the name of this buffer to `human_readable_name`.

File: src/cluster.h [hunk 8]
```diff
@@ -283,14 +288,16 @@ typedef struct {
     char sender[CLUSTER_NAMELEN]; /* Name of the sender node */
     unsigned char myslots[CLUSTER_SLOTS/8];
     char slaveof[CLUSTER_NAMELEN];
-    char myip[NET_IP_STR_LEN];    /* Sender IP, if not all zeroed. */
-    char notused1[32];  /* 32 bytes reserved for future usage. */
+    char myip[NET_IP_STR_LEN];          /* Sender IP, if not all zeroed. */
```
Comment (daniel-house) [inline src/cluster.h]: Why was this line changed?

File: src/cluster.h [hunk 9]
```diff
@@ -283,14 +288,16 @@ typedef struct {
...
+    char hname[CLUSTER_HUMANNAMELEN];   /* Human readable name assigned to the sender */
+    char notused1[32];   /* 32 bytes reserved for future usage. */
```
Comment (daniel-house) [inline src/cluster.h:293]: Why was this line changed?

File: src/redis-cli.c (hunk 2 above)
Comment (daniel-house) [reply src/redis-cli.c]: It appears that both `hname` and `custom_name` are being used, so you shouldn't need these two lines any more.  On the other hand, it looks like they are only used so that you can assign them values.  Can you simply delete these two variables?

File: src/cluster.c [hunk 10]
```diff
@@ -48,6 +48,7 @@ clusterNode *myself = NULL;
 
 clusterNode *createClusterNode(char *nodename, int flags);
 void clusterAddNode(clusterNode *node);
+void setClusterNodeName(clusterNode *node);
```
Comment (daniel-house) [inline src/cluster.c:51]: This is a forward reference.  Generally speaking, this is inferior to simply putting the function definition in the file before any call to the function.  In this case it might be good because it appears to conform to a pre-existing coding style, but it has the drawback that it can defeat compile-and-link-time checking, making it easier for a bug to hide until runtime.

File: src/cluster.c [hunk 11]
```diff
@@ -212,13 +213,17 @@ int clusterLoadConfig(char *filename) {
             n = createClusterNode(argv[0],0);
             clusterAddNode(n);
         }
+
+        /*Human readable name*/
+        memcpy(n->hname, argv[1], strlen(argv[1]) + 1);
```
Comment (daniel-house) [inline src/cluster.c]: What if `strlen(argv[1])` is larger than `sizeof(n->hname)`?

File: src/cluster.c [hunk 12]
```diff
@@ -819,6 +828,25 @@ unsigned int keyHashSlot(char *key, int keylen) {
  * CLUSTER node API
  * -------------------------------------------------------------------------- */
 
+/* Assign a human readable name to nodes for clusters*/
+void setClusterNodeName(clusterNode *node) {
```
Comment (daniel-house) [inline src/cluster.c:832]: Since this function will never be called outside of this file, it would be best to declare it `static`, which hides the function from other files, makes it easier for the compiler to optimize the code, and makes it obvious to human readers that the function will never be called outside of this file.  Unfortunately, static functions are not used anywhere in the Redis code base, so declaring this static would appear to violate the coding standards.

File: src/cluster.c [hunk 13]
```diff
@@ -819,6 +828,25 @@ unsigned int keyHashSlot(char *key, int keylen) {
...
+    if (node->custom_name == 1)
+        return;
+    char ip[strlen(node->ip)];
+    strcpy(ip,node->ip);
```
Comment (daniel-house) [inline src/cluster.c:836]: This will cause a buffer overrun EVERY TIME!

Do you know if `node->ip` is null terminated?

Assuming `node->ip` is null terminated, `strlen(node->ip)` will not count the null terminator, so the `ip` buffer will not be long enough to hold the null terminator and `strcpy` will attempt to write the null terminator after the end of the buffer.

File: src/cluster.h (hunk 9 above)
Comment (hwware) [reply src/cluster.h:293]: To allign with the comments below
